File: 029.DivideTwoIntegers/MarkCSIE/DivideTwoIntegers/src/DivideTwoIntegers.java

public class DivideTwoIntegers {

    public static void main(String[] args) {
        DivideTwoIntegers test = new DivideTwoIntegers();
        System.out.println(test.divide(23, 5));
    }
    
        public int divide(int dividend, int divisor) {
        if (dividend == Integer.MIN_VALUE && divisor == -1) {
            return Integer.MAX_VALUE;
        }

        long lDividend = Math.abs((long) dividend);
        long lDivisor = Math.abs((long) divisor);

        int quotient = 0;
        while (lDividend >= lDivisor) {
            int k = 0;
            while (lDividend >= (lDivisor << k)) {
                k++;
            }
            lDividend -= (lDivisor << (k - 1));
            quotient += 1 << (k - 1);
        }

        if ((dividend > 0 && divisor < 0) || (dividend < 0 && divisor > 0)) {
            return -quotient;
        } else {
            return quotient;
        }
    }
    
}

File: 029.DivideTwoIntegers/wyrj/DivideTwoIntegers.js
var divide = function(dividend, divisor) {
    var MAX_INT = Math.pow(2, 31) - 1;
    var sign = (dividend < 0) ^ (divisor < 0);
    var shifted, multi, result = 0;
    
    dividend = Math.abs(dividend);
    divisor = Math.abs(divisor);
    while(dividend >= divisor) {
        shifted = divisor;
        multi = 1;
        while(dividend >= (shifted << 1) && (shifted << 1) > 0) {
            shifted = shifted << 1;
            multi = multi << 1;
        }
        result += multi;
        dividend -= shifted;
    }
    result = (sign) ? -result : result;
    return (result > MAX_INT) ? MAX_INT : result;
};